### PR TITLE
add option to exclude rabbitmq routingKey from resource name

### PR DESCRIPTION
**Pull request by amarziali**
# What Does This Do

Add an option to exclude rabbitMQ routing key from the publisher's resource name. This can be used to reduce the cardinality of generated spans' resources.

The feature is driven by the system property  `dd.trace.rabbit.include.routingkey.in.resource` (or env `DD_TRACE_RABBIT_INCLUDE_ROUTINGKEY_IN_RESOURCE`) which is enabled by default (it's the current behaviour).

Setting it to false will exclude the routingKey to be part of the span resource name

# Motivation

# Additional Notes

Being a breaking change, the option is disabled by default

### CODE REVIEW

File: dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java [hunk 1]
```diff
@@ -63,6 +63,9 @@ public final class TraceInstrumentationConfig {
   public static final String RABBIT_PROPAGATION_DISABLED_EXCHANGES =
       "rabbit.propagation.disabled.exchanges";
 
+  public static final String RABBIT_EXCLUDE_ROUTINGKEY_FROM_RESOURCE =
+      "rabbit.exclude.routingkey.from.resource";
```
Comment (bantonsson) [inline dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java:67]: This is just a nitpick, but to me it feels better if the setting is worded the other way around, like `rabbit.include.routingkey.in.resource`, with a default of `true`.

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java [hunk 2]
```diff
@@ -149,6 +149,15 @@ private String normalizeQueueName(String queueName) {
     return queueName;
   }
 
+  private String buildResourceName(
+      final String opName, final String exchangeName, final String routingKey) {
+    final String prefix = opName + " " + exchangeName;
+    if (Config.get().isRabbitRoutingKeyExcludedFromResourceName()) {
+      return prefix;
+    }
+    return prefix + " -> " + routingKey;
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java:158]: Please use a `StringBuilder` in this method that can be presized to the max including the routing key.

**Review by bantonsson — COMMENTED** [review]
Sorry for not getting to this earlier.

File: dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java (hunk 1 above)
Comment (amarziali) [reply dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java]: Yep I changed accordingly. I agree this is more straightforward

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java (hunk 2 above)
Comment (amarziali) [reply dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java]: All right I imagined the compiler was already doing some optimization. I allocated a `StringBuilder` with a size like the worst case as suggested

**Comment by sdwa0** [comment]
The system property used in the code is `dd.rabbit.include.routingkey.in.resource` (or env `DD_RABBIT_INCLUDE_ROUTINGKEY_IN_RESOURCE`), which is different than what is mentioned in the description above (there is no `trace.` after `dd.`).

This isn't a big issue, but I just thought mentioning this here in case someone tried the property mentioned in the description and didn't get the desired result.

**Comment by amarziali** [comment]
> 

Thanks for having catched it out. I edited the description accordingly

**Pull request merged**